### PR TITLE
Sensors - Allow changing RRD type in YAML and PHP sensor discovery

### PR DESCRIPTION
**Pull request by PipoCanaja**
Replaces #13854

A new DB column in sensors stores the RRD type (which defaults to GAUGE as it is currently). This is a draft currently. Will update this desc as it reaches a "runnable" status. 

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by murrant** [comment]
Seems reasonable, do any changes need to be made to the graphs?

**Comment by PipoCanaja** [comment]
> Seems reasonable, do any changes need to be made to the graphs?


> Seems reasonable, do any changes need to be made to the graphs?

Not necessarilly. May be some text/graphical changes to add "per second" to the unit, etc. Will see how it goes with tests.

**Comment by PipoCanaja** [comment]
I don't understand why the airos-af-ltu test fails. The diff is as expected and has not impacted wireless-sensors ... 
```
librenms@librenms:~$ git diff origin/master tests/data/airos-af-ltu.json 
diff --git a/tests/data/airos-af-ltu.json b/tests/data/airos-af-ltu.json
index 4308f529bf..520b842188 100644
--- a/tests/data/airos-af-ltu.json
+++ b/tests/data/airos-af-ltu.json
@@ -1297,8 +1297,8 @@
                     "ifName": "br0",
                     "portName": null,
                     "ifIndex": 5,
-                    "ifSpeed": null,
-                    "ifSpeed_prev": 0,
+                    "ifSpeed": 0,
+                    "ifSpeed_prev": null,
                     "ifConnectorPresent": "true",
                     "ifPromiscuousMode": "false",
                     "ifOperStatus": "up",
@@ -1402,8 +1402,8 @@
                     "ifName": "br2",
                     "portName": null,
                     "ifIndex": 6,
-                    "ifSpeed": null,
-                    "ifSpeed_prev": 0,
+                    "ifSpeed": 0,
+                    "ifSpeed_prev": null,
                     "ifConnectorPresent": "true",
                     "ifPromiscuousMode": "false",
                     "ifOperStatus": "down",
@@ -1717,8 +1717,8 @@
                     "ifName": "br0",
                     "portName": null,
                     "ifIndex": 10,
-                    "ifSpeed": null,
-                    "ifSpeed_prev": 0,
+                    "ifSpeed": 0,
+                    "ifSpeed_prev": null,
                     "ifConnectorPresent": "true",
                     "ifPromiscuousMode": "false",
                     "ifOperStatus": "up",
@@ -1822,8 +1822,8 @@
                     "ifName": "br2",
                     "portName": null,
                     "ifIndex": 11,
-                    "ifSpeed": null,
-                    "ifSpeed_prev": 0,
+                    "ifSpeed": 0,
+                    "ifSpeed_prev": null,
                     "ifConnectorPresent": "true",
                     "ifPromiscuousMode": "false",
                     "ifOperStatus": "down",
@@ -2061,6 +2061,7 @@
                     "entPhysicalIndex_measured": null,
                     "sensor_prev": null,
                     "user_func": null,
+                    "rrd_type": "GAUGE",
                     "state_name": null
                 },
                 {
@@ -2085,6 +2086,7 @@
                     "entPhysicalIndex_measured": null,
                     "sensor_prev": null,
                     "user_func": null,
+                    "rrd_type": "GAUGE",
                     "state_name": null
                 },
                 {
@@ -2109,6 +2111,7 @@
                     "entPhysicalIndex_measured": null,
                     "sensor_prev": null,
                     "user_func": null,
+                    "rrd_type": "GAUGE",
                     "state_name": "afLTUgpsDimensions"
                 },
                 {
@@ -2133,6 +2136,7 @@
                     "entPhysicalIndex_measured": null,
                     "sensor_prev": null,
                     "user_func": null,
+                    "rrd_type": "GAUGE",
                     "state_name": "afLTUgpsStatus"
                 },
                 {
@@ -2157,6 +2161,7 @@
                     "entPhysicalIndex_measured": null,
                     "sensor_prev": null,
                     "user_func": null,
+                    "rrd_type": "GAUGE",
                     "state_name": "afLTURole"
                 },
                 {
@@ -2181,6 +2186,7 @@
                     "entPhysicalIndex_measured": null,
                     "sensor_prev": null,
                     "user_func": null,
+                    "rrd_type": "GAUGE",
                     "state_name": "afLTUStaRxRate"
                 },
                 {
@@ -2205,6 +2211,7 @@
                     "entPhysicalIndex_measured": null,
                     "sensor_prev": null,
                     "user_func": null,
+                    "rrd_type": "GAUGE",
                     "state_name": "afLTUStaTxRate"
                 }
             ],
```

**Comment by PipoCanaja** [comment]
@murrant Whatever I do, save-test-data cannot provide me 100% of files that would pass correctly the tests. Even on my own server, with same DB and same settings, they do not succeed. Would you have any suggestions ?

**Comment by PipoCanaja** [comment]
@librenms/reviewers  : tried my best with tests but still fighting to get it right : save-test-data runs as good as it can, and provides a result that makes sense. 
lnms dev:check -o arista-mos runs well on my VM. 
But we have an error here in the PR tests: 
https://github.com/librenms/librenms/runs/8259961537?check_suite_focus=true

**Comment by murrant** [comment]
Sorry this is a nightmare to update :/

**Comment by PipoCanaja** [comment]
Hi @murrant @Jellyfrog 
I am calling for help :)
On the precise test that is currently failing currently, I don't get the issue : 
`tests/data/arista-mos_metamux48-c-0-16.json`

First, this : 
image

And then this:
image

Do you have any tips ? 

Thx

File: LibreNMS/Device/Sensor.php [hunk 1]
```diff
@@ -58,6 +58,7 @@ class Sensor implements DiscoveryModule, PollerModule
     private $low_warn;
     private $entPhysicalIndex;
     private $entPhysicalMeasured;
+    private $rrd_type = 'GAUGE';
```
Comment (PipoCanaja) [inline LibreNMS/Device/Sensor.php:61]: Hi @librenms/reviewers 
How should I handle this PHP stan error : 
```
GitHub Actions
/ PHP Static Analysis (7.3)
Property LibreNMS\Device\Sensor::$rrd_type has no type specified.
```
Thx

**Comment by Jellyfrog** [comment]
It except sensors but doesn't find any for that device.

**Comment by PipoCanaja** [comment]
> It except sensors but doesn't find any for that device.

But looking at the capture, you see that the data is actually found :) That's my problem. It finds it but complains it doesn't ...

**Comment by PipoCanaja** [comment]
Hi @librenms/reviewers 
Looks like it is OK now. Only have to keep up with master each time a JSON file is added or updated. 
Let me know your opinion on this.

**Review by murrant — APPROVED** [review]

**Comment by Jellyfrog** [comment]
Might miss something, but when running ./scripts/save-test-data.php -o apc -v sua750i, it breaks on:

Error polling sensors module for 127.1.6.1. LibreNMS\Exceptions\InvalidRrdTypeException:  is not valid, must be: GAUGE | DERIVE | COUNTER | ABSOLUTE | DCOUNTER | DDERIVE in librenms/LibreNMS/RRD/RrdDefinition.php:122

because the $sensor doesnt have a rrd_type set:
```
array:25 [
  "sensor_id" => 107
  "sensor_deleted" => 0
  "sensor_class" => "state"
  "device_id" => 13
  "poller_type" => "snmp"
  "sensor_oid" => ".1.3.6.1.4.1.318.1.1.1.4.1.1.0"
  "sensor_index" => "0"
  "sensor_type" => "upsBasicOutputStatus"
  "sensor_descr" => "Output Status"
  "group" => null
  "sensor_divisor" => 1
  "sensor_multiplier" => 1
  "sensor_current" => 2.0
  "sensor_limit" => null
  "sensor_limit_warn" => null
  "sensor_limit_low" => null
  "sensor_limit_low_warn" => null
  "sensor_alert" => 1
  "sensor_custom" => "No"
  "entPhysicalIndex" => null
  "entPhysicalIndex_measured" => null
  "lastupdate" => "2022-10-31 18:00:39"
  "sensor_prev" => null
  "user_func" => null
  "new_value" => "2"
]
```

**Comment by PipoCanaja** [comment]
> Might miss something, but when running ./scripts/save-test-data.php -o apc -v sua750i, it breaks on:

did you run the migrations ? I would be the DB is not in the correct state.

**Comment by Jellyfrog** [comment]
Oh. No :) it's probably that then. I'll recheck tomorrow

**Comment by PipoCanaja** [comment]
when I do test generation, I usually run "lnms dev:check -o myOsName" at least up to the "refreshing database ... done". That way I am sure the DB is empty and clean.

**Pull request merged**